org.eclipse.xtext.xbase/src/org/eclipse/xtext/xbase/typesystem/internal/AbstractTypeComputationState.java

```diff
@@ -410,15 +404,59 @@ public LightweightTypeReference getActualType(XExpression expression) {
 		Iterable<IEObjectDescription> descriptions = reentrantTypeResolver.getScopeProviderAccess().getCandidateDescriptions(
 				featureCall, XbasePackage.Literals.XABSTRACT_FEATURE_CALL__FEATURE, proxyOrResolved, featureScopeSession, demandResolvedTypes);
 		List<IFeatureLinkingCandidate> resultList = Lists.newArrayList();
+		AbstractPendingLinkingCandidate<?> previouslyResolved = resolvedTypes.forwardLinking(featureCall);
 		for(IEObjectDescription description: descriptions) {
-			resultList.add(createCandidate(featureCall, demandComputedTypes, toIdentifiableDescription(description)));
+			IFeatureLinkingCandidate candidate = createCandidate(featureCall, demandComputedTypes, toIdentifiableDescription(description));
+			// TODO apply to constructors, too
+			if (matches(previouslyResolved, candidate)) {
+				return Collections.singletonList(candidate);
+			}
+			resultList.add(candidate);
 		}
 		if (resultList.isEmpty()) {
 			resultList.add(new NullFeatureLinkingCandidate(featureCall, this));
 		}
 		return resultList;
 	}
+
+	private boolean matches(AbstractPendingLinkingCandidate<?> previouslyResolved, IFeatureLinkingCandidate candidate) {
+		if (previouslyResolved == null) {
+			return false;
+		}
+		if (!previouslyResolved.getClass().equals(candidate.getClass())) {
+			return false;
+		}
+		AbstractPendingLinkingCandidate<?> casted = (AbstractPendingLinkingCandidate<?>)candidate;
+		IIdentifiableElementDescription prevDescription = previouslyResolved.description;
+		IIdentifiableElementDescription description = casted.description;
+		if (!prevDescription.getShadowingKey().equals(description.getShadowingKey())) {
+			return false;
+		}
+		if (!prevDescription.getElementOrProxy().equals(description.getElementOrProxy())) {
+			return false;
+		}
+		if (!equalTypes(prevDescription.getSyntacticReceiverType(), description.getSyntacticReceiverType())) {
+			return false;
+		}
+		if (!equalTypes(prevDescription.getImplicitReceiverType(), description.getImplicitReceiverType())) {
+			return false;
+		}
+		if (!equalTypes(prevDescription.getImplicitFirstArgumentType(), description.getImplicitFirstArgumentType())) {
+			return false;
+		}
+		return true;
+	}
 	
+	private boolean equalTypes(LightweightTypeReference left, LightweightTypeReference right) {
+		if (left == right) {
+			return true;
+		}
+		if (right == null) {
+			return false;
+		}
+		return left.getUniqueIdentifier().equals(right.getUniqueIdentifier());
```

Comment:
Left is not checked for null. From above implementation this might be a valid assumption. For clean implementation of "equalTypes" I would suggest to add the check anyways (esspecially because "left" and "right" are of same type and can be exchanged easaly ...).

Reply:
Fixed.